### PR TITLE
[Kernels] add necessary kernels (llama & bloom) for attention forward and kv-cache manager 

### PR DESCRIPTION
**Pull request by tiandiao123**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?
This PR is used to start to add kernels and related attention forward for kv-cache manager and non-kv cacha manager 
for BS <= 2, we can use flash-attention-2, rotary embedding and rmsnorm(layernorm) cuda kernels to speed up, when bs > 2, we can use kv-cache manager to increase throughput. 

together work with @yuanheng-zhao @CjhHa1 @yuehuayingxueluo @Xu-Kai 

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/kernel/triton/context_attention.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        m_i = tl.zeros([BLOCK_M], dtype=tl.float32) - float("inf")
+        l_i = tl.zeros([BLOCK_M], dtype=tl.float32)
+        acc = tl.zeros([BLOCK_M, BLOCK_DMODEL], dtype=tl.float32)
+        
+        if alibi_ptr is not None:
+            alibi_m = tl.load(alibi_ptr + cur_head)
+
+        block_mask = tl.where(block_start_loc < cur_batch_seq_len, 1, 0)
+
+        # modify from
```
Comment (Xu-Kai) [inline colossalai/kernel/triton/context_attention.py:65]: The comment is not complete.

File: tests/test_infer_ops/triton/test_bloom_context_attention.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import math
+from packaging import version
+
+import torch
+from torch import nn
+from torch.nn import functional as F
+
+
+from tests.test_kernels.triton.utils import benchmark, torch_context_attention
+from colossalai.kernel.triton.context_attention import bloom_context_attn_fwd
```
Comment (Xu-Kai) [inline tests/test_infer_ops/triton/test_bloom_context_attention.py:11]: May be put in try block? If there is no triton, function can not be imported.

File: colossalai/kernel/triton/context_attention.py (hunk 1 above)
Comment (tiandiao123) [reply colossalai/kernel/triton/context_attention.py:65]: comments has been added to the top of this function.

File: tests/test_infer_ops/triton/test_copy_kv_dest.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import pytest
+from packaging import version
+
+import torch
+from torch import nn
+
+from tests.test_kernels.triton.utils import benchmark
+from colossalai.kernel.triton.copy_kv_cache_dest import copy_kv_cache_to_dest
```
Comment (Xu-Kai) [inline tests/test_infer_ops/triton/test_copy_kv_dest.py:8]: This can put in try block.

File: tests/test_infer_ops/triton/test_self_attention_nonfusion.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 from torch import nn 
 import torch.nn.functional as F
 
-from colossalai.kernel.triton.ops import self_attention_compute_using_triton
+from colossalai.kernel.triton.self_attention_nofusion import self_attention_compute_using_triton
```
Comment (Xu-Kai) [inline tests/test_infer_ops/triton/test_self_attention_nonfusion.py:7]: Same as before.

File: colossalai/kernel/triton/softmax.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        block_size = max(triton.next_power_of_2(num_cols), 2)
+        num_warps = 16
+        if block_size >= 4096:
+            num_warps = 16
+        elif block_size >= 2048:
+            num_warps = 8
+        else:
+            num_warps = 4
+
+        if num_rows <= 350000:
```
Comment (Xu-Kai) [inline colossalai/kernel/triton/softmax.py:70]: What does 350000 mean?

Comment (Xu-Kai) [reply colossalai/kernel/triton/softmax.py:70]: To consitrain the number of grid.

**Review by Xu-Kai — APPROVED** [review]

**Review by Xu-Kai — APPROVED** [review]

**Review by Xu-Kai — APPROVED** [review]

**Review by Xu-Kai — APPROVED** [review]

**Review by yuanheng-zhao — APPROVED** [review]
LGTM
